ocdata/core/surface.py

```diff
@@ -0,0 +1,275 @@
+
+    def tag_surface_atoms(self):
+        """
+        Sets the tags of an `ase.Atoms` object. Any atom that we consider a "bulk"
+        atom will have a tag of 0, and any atom that we consider a "surface" atom
+        will have a tag of 1. We use a combination of Voronoi neighbor algorithms
+        (adapted from from `pymatgen.core.surface.Slab.get_surface_sites`; see
+        https://pymatgen.org/pymatgen.core.surface.html) and a distance cutoff.
+
+        Arg:
+            bulk_atoms      `ase.Atoms` format of the respective bulk structure
+            surface_atoms   The surface where you are trying to find surface sites in
+                            `ase.Atoms` format
+        """
+        voronoi_tags = find_surface_atoms_with_voronoi(self.bulk.atoms, self.atoms)
+        height_tags = find_surface_atoms_by_height(self.atoms)
+        # If either of the methods consider an atom a "surface atom", then tag it as such.
+        tags = [max(v_tag, h_tag) for v_tag, h_tag in zip(voronoi_tags, height_tags)]
+        self.atoms.set_tags(tags)
+
+    def set_fixed_atom_constraints(self):
+        """
+        This function fixes sub-surface atoms of a surface. Also works on systems
+        that have surface + adsorbate(s), as long as the bulk atoms are tagged with
+        `0`, surface atoms are tagged with `1`, and the adsorbate atoms are tagged
+        with `2` or above.
+
+        This function is used for both surface atoms and the combined surface+adsorbate
+
+        Inputs:
+            atoms           `ase.Atoms` class of the surface system. The tags of
+                            these atoms must be set such that any bulk/surface
+                            atoms are tagged with `0` or `1`, resectively, and any
+                            adsorbate atom is tagged with a 2 or above.
+        Returns:
+            atoms           A deep copy of the `atoms` argument, but where the appropriate
+                            atoms are constrained.
+        """
+        # We'll be making a `mask` list to feed to the `FixAtoms` class. This
+        # list should contain a `True` if we want an atom to be constrained, and
+        # `False` otherwise.
+        mask = [True if atom.tag == 0 else False for atom in self.atoms]
+        self.atoms.constraints += [FixAtoms(mask=mask)]
+
+    def __len__(self):
+        return len(self.atoms)
+
+    def __str__(self):
+        return f"Surface: {self.atoms.get_chemical_formula()}"
+
+    def __repr__(self):
+        return self.__str__()
+
+
+def tile_atoms(atoms, min_xy=MIN_XY):
+    """
+    This function will repeat an atoms structure in the x and y direction until
+    the x and y dimensions are at least as wide as the MIN_XY constant.
+
+    Arguments
+    ---------
+    atoms: ase.Atoms
+        The structure to tile.
+
+    Returns
+    -------
+    atoms_tiled: ase.Atoms
+        The tiled structure.
+    """
+    x_length = np.linalg.norm(atoms.cell[0])
```

Comment:
`cell[0]` `cell[1]` are usually not orthogonal so this naming is a little deceptive